pkg/adhoc/server/server.go

```diff
@@ -0,0 +1,119 @@
+package server
+
+import (
+	"crypto/sha256"
+	"encoding/json"
+	"fmt"
+	"io"
+	"io/fs"
+	"net/http"
+	"os"
+	"path/filepath"
+	"sync"
+	"time"
+
+	"github.com/gorilla/mux"
+	"github.com/sirupsen/logrus"
+
+	"github.com/pyroscope-io/pyroscope/pkg/adhoc/util"
+)
+
+type profile struct {
+	ID        string    `json:"id"`
+	Name      string    `json:"name"`
+	UpdatedAt time.Time `json:"updated_at"`
+}
+
+type server struct {
+	log      logrus.FieldLogger
+	enabled  bool
+	profiles map[string]profile
+	mtx      sync.RWMutex
+}
+
+type Server interface {
+	AddRoutes(router *mux.Router) http.HandlerFunc
+}
+
+func New(log logrus.FieldLogger, enabled bool) Server {
+	return &server{log: log, enabled: enabled, profiles: make(map[string]profile)}
+}
+
+func (s *server) AddRoutes(r *mux.Router) http.HandlerFunc {
+	if s.enabled {
+		r.HandleFunc("/v1/profiles", s.Profiles)
+		r.HandleFunc("/v1/profile/{id:[0-9a-f]+}", s.Profile)
+		r.HandleFunc("/v1/diff/{id1[0-9a-f]+}/{id2[0-9a-f]+}", s.Diff)
+	}
+	return r.ServeHTTP
+}
+
+// Profiles retrieves the list of profiles for the local pyroscope data directory.
+// The profiles are assigned a unique ID (hash based) which is then used for retrieval.
+// This requires a bit of extra work to setup the IDs but prevents
+// the clients from accesing the filesystem directly, removing that whole attack vector.
+//
+// The profiles are retrieved every time the endpoint is requested,
+// which should be good enough as massive access to this auth endpoint is not expected.
+func (s *server) Profiles(w http.ResponseWriter, _ *http.Request) {
+	profiles := make(map[string]profile, 0)
+	err := filepath.Walk(util.DataDirectory(), func(path string, info fs.FileInfo, err error) error {
+		if err != nil {
+			return err
+		}
+		if info.Mode().IsRegular() {
```

Comment:
Maybe we should check if the file is valid. For example I created a `foo` file in the $PYROSCOPE_HOME_DIR

![image](https://user-images.githubusercontent.com/6951209/146552181-7af65e32-14ed-4e77-8ea5-b49a8337f586.png)

Reply:
Added some basic server-side validation in https://github.com/pyroscope-io/pyroscope/pull/624/commits/dea2544a7413afd5f5b0be67662fe08dc9dd8d42